File: src/main/java/com/machfour/macros/cli/MealSpec.java
package com.machfour.macros.cli;

import com.machfour.macros.objects.Meal;
import com.machfour.macros.storage.MacrosDatabase;
import com.machfour.macros.util.DateStamp;
import org.jetbrains.annotations.NotNull;

import java.sql.SQLException;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

import static com.machfour.macros.cli.ArgParsing.Status.OPT_ARG_MISSING;
import static com.machfour.macros.cli.ArgParsing.Status.ARG_FOUND;

/*
 * Stores information for a meal being input (specified) via text fields,
 * and implements default behaviour for retrieving the meal when not
 * all fields are fully specified
 */
class MealSpec {

    private final String name;
    private final DateStamp day;
    // whether parameters were actually given (true) or the default value used (false)
    private final boolean mealSpecified;
    private final boolean daySpecified;
    private Meal processedObject;
    // whether the processedObject was newly created or already existed in the DB
    private boolean created;
    private String error;

    private MealSpec(String name, DateStamp day) {
        this.name = name;
        this.day = day;
        this.mealSpecified = (name != null);
        this.daySpecified = (day != null);
    }

    MealSpec(ArgParsing.Result dayArg, ArgParsing.Result mealArg) {
        DateStamp day = ArgParsing.dayStringParse(dayArg.argument());
        if (day == null) {
            error = String.format("Invalid day format: '%s'. ", dayArg.argument());
            error += "Must be a number (e.g. 0 for today, -1 for yesterday), or a date: yyyy-mm-dd";
        } else if (dayArg.status() == OPT_ARG_MISSING) {
            error = "-d option requires an argument: <day>";
        }
        this.day = day;
        this.daySpecified = (day != null) && dayArg.status() == ARG_FOUND;

        String mealName = mealArg.argument();
        this.name = mealName;
        this.mealSpecified = mealName != null && mealArg.status() == ARG_FOUND;
        if (mealArg.status() == OPT_ARG_MISSING) {
            error = "-m option requires an argument: <meal>";
        }
    }


    void processMealSpec(MacrosDatabase db, boolean create) {
        if (error != null) {
            // skip processing if there are already errors
            return;
        }
        // cases:
        // no meal specified -> use current meal (exists)
        // no meal specified -> no meal exists
        // meal specified that exists -> use it
        // meal specified that does not exist -> create it
        Map<String, Meal> mealsForDay;
        try {
            mealsForDay = db.getMealsForDay(day);
        } catch (SQLException e) {
            error = String.format("Error retrieving meals for day %s: %s", day.toString(), e.getMessage());
            return;
        }
        if (!mealSpecified) {
            if (!mealsForDay.isEmpty()) {
                // use most recently modified meal today
                processedObject = Collections.max(mealsForDay.values(), Comparator.comparingLong(Meal::modifyTime));
            } else {
                error = "No meals recorded on " + CliUtils.prettyDay(day);
            }
        } else if (mealsForDay.containsKey(name)) {
            processedObject = mealsForDay.get(name);
            created = false;
        } else if (create) {
            try {
                processedObject = db.getOrCreateMeal(day, name);
                created = true;
            } catch (SQLException e) {
                error = "Error retrieving meal: " + e.getMessage();
                return;
            }
        } else {
            // meal doesn't exist and not allowed to create new meal
            error = String.format("No meal with name '%s' found on %s", name, CliUtils.prettyDay(day));
        }
        assert (error != null || processedObject != null) : "No error message but no created object";
        if (error != null) {
            return;
        }
    }


    // extracts a meal specification from the argument list using the following rules:
    // -d <day> specifies a day to search for, or on which to create the meal if create = true
    // -m <name> specifies a name for the meal, which is created if create = true and it doesn't already exist.
    // Both options can be omitted under certain condititions:
    // If -d is omitted then the current day is used.
    // If there are no meals recorded for the day, then an error is given.
    @NotNull
    static MealSpec makeMealSpec(List<String> args) {
        ArgParsing.Result dayArg = ArgParsing.findArgument(args, "-d");
        ArgParsing.Result mealArg = ArgParsing.findArgument(args, "-m");
        return new MealSpec(dayArg, mealArg);
    }
    @NotNull
    static MealSpec makeMealSpec(String name, DateStamp day) {
        return new MealSpec(name, day);
    }

    @NotNull
    static MealSpec makeMealSpec(ArgParsing.Result nameArg, ArgParsing.Result dayArg) {
        return new MealSpec(nameArg, dayArg);
    }
    @NotNull
    static MealSpec makeMealSpec(String name) {
        return new MealSpec(name, DateStamp.forCurrentDate());
    }

    String name() {
        return name;
    }

    DateStamp day() {
        return day;
    }

    boolean mealSpecified() {
        return mealSpecified;
    }

    boolean daySpecified() {
        return daySpecified;
    }

    Meal processedObject() {
        return processedObject;
    }

    boolean created() {
        return created;
    }

    String error() {
        return error;
    }
}

File: libmacros/src/main/kotlin/com/machfour/macros/parsing/MealSpec.kt
package com.machfour.macros.parsing

import com.machfour.datestamp.DateStamp
import com.machfour.macros.core.EntityId
import com.machfour.macros.core.ObjectSource
import com.machfour.macros.entities.Meal
import com.machfour.macros.queries.getMealForDayWithName
import com.machfour.macros.queries.getMealsForDay
import com.machfour.macros.queries.saveObject
import com.machfour.macros.schema.MealTable
import com.machfour.macros.sql.SqlDatabase
import com.machfour.macros.sql.SqlException
import com.machfour.macros.sql.rowdata.RowData

/*
 * Stores information for a meal being input (specified) via text fields,
 * and implements default behaviour for retrieving the meal when not
 * all fields are fully specified
 */
class MealSpec(val name: String?, val day: DateStamp?, error: String? = null) {

    // whether parameters were actually given (true) or the default value used (false)
    val isMealSpecified: Boolean = name != null

    var processedObject: Meal? = null
        private set

    // whether the processedObject was newly created or already existed in the DB
    var isCreated = false
        private set

    var error: String? = error
        private set

    // has process() been called?
    private var processed = false

    @Throws(SqlException::class)
    private fun getOrCreateMeal(ds: SqlDatabase, day: DateStamp, name: String): Meal {
        // if it already exists return it
        val matchingMeal = getMealForDayWithName(ds, day, name)
        if (matchingMeal != null) {
            return matchingMeal
        }

        // else create a new meal, save and return it
        val newMeal = RowData(MealTable).run {
            put(MealTable.DAY, day)
            put(MealTable.NAME, name)
            Meal.factory.construct(this, ObjectSource.USER_NEW)
        }

        saveObject(ds, Meal.factory,newMeal)
        // get it back again, so that it has an ID and stuff
        val newlySavedMeal = getMealForDayWithName(ds, day, name)
        check(newlySavedMeal != null) { "Couldn't find newly saved meal in day $day" }
        return newlySavedMeal
    }

    fun process(ds: SqlDatabase, create: Boolean) {
        if (processed || error != null) {
            // skip processing if there are already errors
            return
        }
        processed = true // only let process() be called once

        if (day == null) {
            throw IllegalStateException("Day cannot be null once process() is called")
        }

        // cases:
        // no meal specified -> use current meal (exists)
        // no meal specified -> no meal exists
        // meal specified that exists -> use it
        // meal specified that does not exist -> create it
        val mealsForDay: Map<EntityId, Meal>
        try {
            mealsForDay = getMealsForDay(ds, day)
        } catch (e: SqlException) {
            error = "Error retrieving meals for day ${day}: $e"
            return
        }
        if (!isMealSpecified) {
            if (mealsForDay.isNotEmpty()) {
                // use most recently modified meal today
                processedObject = mealsForDay.values.maxByOrNull { it.modifyTime }!!
            } else {
                error = "No meals recorded on " + day.prettyPrint()
            }
        } else {
            // if a name was given, try to find a matching meal with that name
            // name cannot be null since it is implied by isMealSpecified
            val name = name!!
            val nameMatch = mealsForDay.values.firstOrNull { it.name == name }
            when {
                nameMatch != null -> {
                    processedObject = nameMatch
                    isCreated = false
                }
                create -> {
                    try {
                        processedObject = getOrCreateMeal(ds, day, name)
                        isCreated = true
                    } catch (e: SqlException) {
                        error = "Error retrieving meal: " + e.message
                        return
                    }
                }
                else -> {
                    // meal doesn't exist and not allowed to create new meal
                    error = "No meal with name '$name' found on ${day.prettyPrint()}"
                }
            }
        }
        check(error != null || processedObject != null) { "No error message but no created object" }
        if (error != null) {
            return
        }
    }


}